workspace/update-injected-packages/src/DirPatcher.ts

```diff
@@ -0,0 +1,208 @@
+ */
+export async function applyDiff (optimizedDirDiff: DirDiff, sourceDir: string, targetDir: string): Promise<void> {
+  async function addRecursive (sourcePath: string, targetPath: string, value: Value): Promise<void> {
+    const makeParent = () => fs.mkdirSync(path.dirname(targetPath), { recursive: true })
+    if (value === DIR) {
+      await fs.promises.mkdir(targetPath, { recursive: true })
+    } else if (typeof value === 'string') {
+      makeParent()
+      const symlinkTarget = value
+      if (process.platform === 'win32') {
+        const symlinkRealTarget = path.resolve(sourceDir, symlinkTarget)
+        await fs.promises.symlink(symlinkRealTarget, targetPath, 'junction')
+      } else {
+        await fs.promises.symlink(symlinkTarget, targetPath)
+      }
+    } else if (typeof value === 'number') {
+      makeParent()
+      await fs.promises.link(sourcePath, targetPath)
+    } else {
+      // eslint-disable-next-line @typescript-eslint/no-unused-vars
+      const _: never = value // static type guard
+    }
+  }
+
+  async function removeRecursive (targetPath: string): Promise<void> {
+    try {
+      await fs.promises.rm(targetPath, { recursive: true, force: true })
+    } catch (error) {
+      if (!util.types.isNativeError(error) || !('code' in error) || (error.code !== 'ENOENT')) {
+        throw error
+      }
+    }
+  }
+
+  const adding = Promise.all(optimizedDirDiff.added.map(async item => {
+    const sourcePath = path.join(sourceDir, item.path)
+    const targetPath = path.join(targetDir, item.path)
+    await addRecursive(sourcePath, targetPath, item.newValue)
+  }))
+
+  const removing = Promise.all(optimizedDirDiff.removed.map(async item => {
+    const targetPath = path.join(targetDir, item.path)
+    await removeRecursive(targetPath)
+  }))
+
+  const modifying = Promise.all(optimizedDirDiff.modified.map(async item => {
+    const sourcePath = path.join(sourceDir, item.path)
+    const targetPath = path.join(targetDir, item.path)
+    if (item.oldValue === item.newValue) return
+    await removeRecursive(targetPath)
+    await addRecursive(sourcePath, targetPath, item.newValue)
+  }))
+
+  await Promise.all([adding, removing, modifying])
+}
+
+/**
+ * Convert a files map, which is a map from relative path of each file to their real paths,
+ * into an inodes map, which is a map from relative path of every file and directory to their inode type.
+ */
+export async function extendFilesMap (filesMap: Record<string, string>): Promise<InodeMap> {
+  const result: InodeMap = {
+    '.': DIR,
+  }
+
+  function addInodeAndAncestors (relativePath: string, value: Value): void {
+    if (relativePath && relativePath !== '.' && !result[relativePath]) {
+      result[relativePath] = value
+      addInodeAndAncestors(path.dirname(relativePath), DIR)
+    }
+  }
+
+  await Promise.all(Object.entries(filesMap).map(async ([relativePath, realPath]) => {
+    const stats = await fs.promises.lstat(realPath)
```

Comment:
fetchFromDir already reads lstat. So it isn't well optimized to read the stat again for each file.

Reply:
I have created https://github.com/pnpm/pnpm/pull/9128. The reason it was in a separate branch is because `lstat` is only really called in `realFileStat` which requires `resolveSymlinks: true` which I wasn't sure whether it should be enabled.

Reply:
https://github.com/pnpm/pnpm/pull/9100/commits/62a0af2061cda43f06a9052d50beb6f215f4b912 will make it reuse the results of `fs.stat` (instead of `fs.lstat`) from `@pnpm/directory-fetcher`.

This of course would be buggy with symlinks. But since you said they are not important, I will ignore it.